File: Go/src/sudoku/sudokusolver.go
package sudoku

import (
	"sync"
)

func Solve(sudoku Sudoku) Sudoku {

	// Firstly, we fill the cell with just one candidate number
	sudoku.MapAndReduce()

	sudokuOut := sudoku.Copy()

	// Checking if MapAndReduce was sufficient to solve it
	if sudokuOut.IsSolved() {
		return sudokuOut
	}

	// Making a list of cell to be updated
	cellList := make([]Cell, 0)
	for rowPos, row := range sudokuOut {

		for colPos, num := range row {

			if num == 0 {
				cell := sudokuOut.setCandidateNumbers(rowPos, colPos)
				cellList = append(cellList, cell)
			}

		}
	}

	// Calling the recursive function that solve the sudoku
	solution, _ := solvePar(sudokuOut, cellList)

	return solution
}

// Solve the sudoku in a concurrent way. It takes a sudoku board and the
// list of unfilled cell as input
func solvePar(sudoku Sudoku, cells []Cell) (Sudoku, bool) {

	// At every recursive call, I reduce cells
	if len(cells) == 0 {
		return sudoku, sudoku.IsSolved()
	}

	// Takes the cell to evaluate
	cellToEvaluate := cells[0]

	// Creates a channel where to collect all the results
	chanSudokuSolve := make(chan Channel)
	// Creates a wait group that waits for every goroutine to finish
	wg := new(sync.WaitGroup)

	// For every candidate number with value true, update the sudoku
	// So every time, it generates multiple incomplete solutions that run concurrently
	for num, b := range cellToEvaluate.CandidateNumbers {

		wg.Add(1)

		// Call the solve function recursively, but as a go routine thread so that it executes asynchronously
		go func(sudokuIn Sudoku, cell Cell, fillVal int, val bool, wg *sync.WaitGroup, c *chan Channel) {

			defer wg.Done()

			if val {

				sudokuOut := sudoku.Copy()

				sudokuOut.Update(cell, fillVal)

				newCellList := make([]Cell, len(cells)-1) // I'm removing the head of the list because at
				copy(newCellList, cells[1:])              // the next call, I don't need it anymore

				// Recursice call with a new shorter list of cell to complete
				solutionInter, solved := solvePar(sudokuOut, newCellList)

				*c <- Channel{Intermediate: solutionInter, Solved: solved}

			}
		}(sudoku, cellToEvaluate, num, b, wg, &chanSudokuSolve)
	}

	// Wait until every routine complited and close the channel
	go func(wg *sync.WaitGroup, c chan Channel) {
		wg.Wait()
		close(c)
	}(wg, chanSudokuSolve)

	// Search for the solution: traverse the channel until it finds a valid solution
	for sol := range chanSudokuSolve {

		if sol.Solved {
			return sol.Intermediate, sol.Solved
		}
	}

	return sudoku, sudoku.IsSolved()
}

File: Erlang/src/sudoku/sudoku_solver.erl
-module(sudoku_solver).

%% API
-export([solve/4]).

solve(SudokuBoard, {Row, Col}, Value, Father) ->

  % Update the cell
  IntSudoku = sudoku:set_cell(SudokuBoard, Row, Col, Value),
  
  % Check whether it's solved or not
  case sudoku:is_solved(IntSudoku) of
    true ->
      Father ! {solved, IntSudoku, self()};
    false ->
      case sudoku:is_valid(IntSudoku) of
        false ->
          Father ! {not_valid, self()};
        true ->
          continue(IntSudoku, Father)
      end
  end.

continue(SudokuBoard, Father) ->

  % Update the board
  InitialSudoku = sudoku:init(SudokuBoard),
  SudokuReduced = sudoku:map_and_reduce(InitialSudoku),

  % Check whether it's solved or not
  case sudoku:is_solved(SudokuReduced) of
    true ->
      Father ! {solved, SudokuReduced, self()};
    false ->
      case sudoku:is_valid(SudokuReduced) of
        false ->
          Father ! {not_valid, self()};
        true ->
          % Search the first empty cell and get its candidate numbers 
          {Row, Col} = sudoku:search_first_empty(SudokuReduced),
          Candidates = maps:keys(sudoku:get_candidate(SudokuReduced, Row, Col)),

          % For any candidate number, start a process 
          _PidS = lists:map(fun(Num) ->
            spawn_link(sudoku_solver, solve, [SudokuReduced, {Row, Col}, Num, Father])
                            end,Candidates)
      end
  end.




